### PR TITLE
Take one bucket, block its public access

### PR DESCRIPTION
**Pull request by NovemberTang**
## What does this change?

An early implementation of the script. It takes in

- bucket name (required)
- AWS profile (optional)
- AWS region (optional)

and attempts to block all public access for the named bucket

Eventually, I would like this to evolve into a project that performs this operation on all buckets in an account, apart from named exceptions, but this is a good starting point.

## Why Go?

I wanted this project to be available as a small binary that developers could download directly from GitHub. Also, I wanted to experiment with Go a bit more, and this small, temporary, standalone project seemed like a good time to do it.

## How to test

Create a bucket in S3, and run this script against it.

## How can we measure success?

This project exists because we are required to comply with AWS Foundational Security Best Practices (FSBP) control S3.8, which requires that each bucket has its own policy blocking all public access. Teams will be able to run this script, and the number of buckets failing S3.8 should fall

## Have we considered potential risks?

This may block access to buckets that need to be public. Teams should try to ensure that this operation is safe, and doublecheck any downstream services

## Output

```
Credentials validated for profile: <AWS_PROFILE>
Found bucket: <BUCKET_NAME> in region: eu-west-1
Public access blocked for bucket: <BUCKET_NAME>
```

### CODE REVIEW

File: src/bucketblocker/main.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"flag"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/sts"
+)
+
+func blockPublicAccess(name string, s3Client s3.S3) (*s3.PutPublicAccessBlockOutput, error) {
```
Comment (marsavar) [inline src/bucketblocker/main.go:14]: Generally speaking I think big structures (such as clients) should be passed by ref rather than value, i.e. `s3Client *s3.S3`. 
Also a personal pet peeve, but I think the client should be the first argument

File: src/bucketblocker/main.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		},
+	})
+	if err != nil {
+		return resp, err
+	}
+	fmt.Println("Public access blocked for bucket: " + name)
+	return resp, nil
+}
+
+func validateCredentials(stsClient sts.STS, profile string) (*sts.GetCallerIdentityOutput, error) {
```
Comment (marsavar) [inline src/bucketblocker/main.go:31]: Same here
```suggestion
func validateCredentials(stsClient *sts.STS, profile string) (*sts.GetCallerIdentityOutput, error) {
```
Same as above

File: src/bucketblocker/main.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	return resp, nil
+}
+
+func main() {
+	name := flag.String("bucket", "", "The name of the bucket to block")
+	profile := flag.String("profile", "default", "The name of the profile to use")
+	region := flag.String("region", "eu-west-1", "The region of the bucket")
+	flag.Parse()
+
```
Comment (marsavar) [inline src/bucketblocker/main.go:44]: Could probably fail early here if the bucket name is empty? Otherwise we're making a bunch of calls to AWS for no reason.

**Review by marsavar — COMMENTED** [review]
Added some thoughts on the code itself, with the caveat that I know very little Go.

My main thought, however, is... I'm not convinced in the the value of maintaining code that essentially wraps the AWS CLI, which can already perform this operation in one line:

```bash
aws s3 put-public-access-block --bucket <name> --region <region> --profile <profile>
```
The AWS CLI also already provides good error handling and I think everyone in P&E has already installed it so there's no need to install go or create workflows to release a versioned/tagged binary. 

TLDR Unless there's a plan to expand this tool (for example by blocking public access to multiple buckets in bulk), the AWS CLI is probably fine?

**Comment by NovemberTang** [comment]
> TLDR Unless there's a plan to expand this tool (for example by blocking public access to multiple buckets in bulk), the AWS CLI is probably fine?

We definitely do want to expand this tool! Ultimately the idea would be for teams to pass in a list of buckets to exclude, and we would change the access settings on all buckets except for the ones listed. I'm just building the tool out incrementally, and this was the smallest piece of coherent work I could come up with

File: README.md [hunk 4]
```diff
@@ -1 +1,20 @@
...
\ No newline at end of file
+# bucket-blocker
+
+## How to run:
+
+Bucket blocker takes up to 3 flags:
+
+- **bucket**: _Required argument_.The name of the bucket to block
+
+- **profile**: _Optional argument_. The profile to use when connecting to AWS. Defaults to `default`
```
[the remark this answers is not in the document]
Comment (NovemberTang) [reply README.md:9]: Alongside `.aws/credentials`, there is an `.aws/config` file where you can set things like a default profile & region, so there's often no need to set these values explicitly.

[the remark this answers is not in the document]
Comment (NovemberTang) [reply README.md:9]: maybe I should set the region to take the default value too...

[the remark this answers is not in the document]
Comment (NovemberTang) [reply README.md:9]: ah wait passing in default doesn't even work actually, it just looks for a named profile called default!

ok explicit it is!
